TypedSyntax/src/vscode.jl

```diff
@@ -0,0 +1,80 @@
+isvscode() = isdefined(Main, :VSCodeServer) && Main.VSCodeServer isa Module
+inlay_hints_available() = isvscode() && isdefined(Main.VSCodeServer, :INLAY_HINTS_ENABLED)
+
+function clearall()
+    if inlay_hints_available()
+        display(Main.VSCodeServer.InlineDisplay(false),  Dict{String, Vector{TypedSyntax.InlayHint}}())
+    end
+    if isvscode()
+        display(Main.VSCodeServer.InlineDisplay(false), TypedSyntax.WarnUnstable[])
+    end
+end
+
+struct WarnUnstable
+    path::String
+    line::Int
+    severity::Int # 0: Error, 1: Warning, 2: Information, 3: Hint
+end
+to_vscode_type(x::WarnUnstable) = (msg="Unstable Type", path = x.path, line = x.line, severity = x.severity)
+function Base.show(io::IO, ::MIME"application/vnd.julia-vscode.diagnostics", warn_diagnostics::AbstractVector{WarnUnstable})
+    return (
+        source = "Cthulhu",
+        items = to_vscode_type.(warn_diagnostics),
+    )
+end
+function add_diagnostic!(warn_diagnostics, node, position, severity)
+    file_path = node.filename
+    line, column = source_location(node, position)
+    push!(warn_diagnostics, WarnUnstable(file_path, line, severity))
+end
+
+const InlayHintKinds = (
+    Type = 1,
+    Parameter = 2
+)
+struct InlayHint
+    line::Int
+    column::Int
+    label::String
+    kind::Union{Nothing, Int}
+end
+to_vscode_type(x::InlayHint) = (position=(x.line, x.column), label=x.label, kind=x.kind)
+function Base.show(io::IO, ::MIME"application/vnd.julia-vscode.inlayHints", type_hints_by_file::AbstractDict{T, <:AbstractVector{InlayHint}}) where T
+    if inlay_hints_available()
+        return Dict{T, Vector{NamedTuple{(:position, :label, :kind), Tuple{Tuple{Int, Int}, String, Union{Nothing, Int}}}}}(
+            filepath => to_vscode_type.(type_hints) for (filepath, type_hints) in type_hints_by_file
+        )
+    end
+    return nothing
+end
+function add_hint!(type_hints, message, node, position; kind=InlayHintKinds.Type)
+    filepath = node.filename
+    line, column = source_location(node, position)
+
+    if filepath ∉ keys(type_hints)
+        type_hints[filepath] = InlayHint[]
+    end
+    push!(type_hints[filepath], InlayHint(line-1, column, message, kind))
+end
+
+function show_annotation(io, @nospecialize(T), post, node, position, type_hints, warn_diagnostics; iswarn::Bool)
```

Comment:
Here too I think getting `type_hints` and `warn_diagnostics` from `io` makes more sense, so you don't need very much vscode-specific code.